### PR TITLE
VEN-808 | Browser tests

### PR DESCRIPTION
**Pull request by TatuArvela**
## Description :sparkles:

Added TestCafe browser tests for
* Navigation and language switching
* Form filling (without sending)
  * New berth application, registered boat, private customer
  * Winter storage application, registered boat on trailer, private customer
  * Unmarked winter storage notice, registered boat on trailer, private customer

Note: Missing cron job from Travis

### Closes :no_good_woman:

* [VEN-808](https://helsinkisolutionoffice.atlassian.net/browse/VEN-808): Implement happy path E2E tests for customer UI

### Automated tests :gear:️

* Browser tests can be executed locally by first running the `start` npm script, and then running the `browser-test:local` npm script in another terminal

### CODE REVIEW

File: browser-tests/selectors/berth.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { screen, within } from '@testing-library/testcafe';
+import { Selector } from 'testcafe';
+
+const selectHarbors = {
```
Comment (santtul) [inline browser-tests/selectors/berth.ts:4]: could we have separate export for this?

File: browser-tests/berth.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+  await t
+    .click(harborListTab)
+    .click(getSelectButtonForHarbor(testData.choice1))
+    .click(getSelectButtonForHarbor(testData.choice2));
+
+  await t.click(nextButton);
+};
+
+const boatInformation = async (t: TestController) => {
```
Comment (santtul) [inline browser-tests/berth.ts:83]: Could this be `fillBoatInformation`?

File: browser-tests/utils/sharedTests.ts [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    .click(boatLength)
+    .pressKey('ctrl+a delete') // ditto for length
+    .typeText(boatLength, testData.boatLength)
+    .typeText(boatName, testData.boatName)
+    .typeText(boatModel, testData.boatModel);
+
+  await t.click(nextButton);
+};
+
+export const applicantInformation = async (
```
Comment (santtul) [inline browser-tests/utils/sharedTests.ts:70]: Could this be `fillApplicantInformation` ?

File: browser-tests/utils/sharedTests.ts [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    .typeText(postalCode, testData.postalCode)
+    .click(municipalitySelect)
+    .click(municipalitySelect.find('option').withText(testData.municipality))
+    .typeText(phoneNumber, testData.phoneNumber)
+    .typeText(emailAddress, testData.emailAddress);
+
+  await t.click(nextButton);
+};
+
+export const overview = async (
```
Comment (santtul) [inline browser-tests/utils/sharedTests.ts:112]: Could this be `assertOverview` or something to have an idea what it's doing?

File: browser-tests/frontPage.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  await t.click(frontPage.unmarkedWinterStorage);
+  await isUnmarkedWinterStoragePage();
+  await navigateToFrontPage();
+});
+
+test('Footer', async (t) => {
+  // 'Venepaikat' Link
+  await t.expect(footer.serviceLink.visible).ok();
+
+  // Link list
```
Comment (santtul) [inline browser-tests/frontPage.ts:72]: I think it's not worth to test static content like this.

File: browser-tests/unmarkedWinterStorage.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    `Malli: ${testData.boatModel}`,
+    `Leveys: ${testData.boatWidth}m`,
+    `Pituus: ${testData.boatLength}m`,
+    `Säilytystapa: Säilytän veneen trailerilla`,
+    `Trailerin rekisterinumero: ${testData.trailerRegistrationNumber}`,
+  ].join('\n');
+  await confirmation(t, expectedBoatInfo);
+});
+
+const areaSelection = async (t: TestController) => {
```
Comment (santtul) [inline browser-tests/unmarkedWinterStorage.ts:53]: `selectArea` ?

File: browser-tests/utils/sharedTests.ts [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+  await t
+    .expect(getHarborHeading(`1. ${testData.choice2}`).exists)
+    .ok()
+    .expect(getHarborHeading(`2. ${testData.choice1}`).exists)
+    .ok();
+
+  await t.click(nextButton);
+};
+
+export const wsBoatInformation = async (
```
Comment (santtul) [inline browser-tests/utils/sharedTests.ts:22]: `fillWsBoatInformation` ?

File: browser-tests/unmarkedWinterStorage.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  const { winterStorageAreaSelect, continueButton } = unmarkedWinterStorage.areaSelection;
+
+  await t
+    .click(winterStorageAreaSelect)
+    .click(winterStorageAreaSelect.find('option').withText(testData.winterStorageArea));
+
+  await t.wait(500).click(continueButton);
+};
+
+const confirmation = async (t: TestController, expectedBoatInfo: string) => {
```
Comment (santtul) [inline browser-tests/unmarkedWinterStorage.ts:63]: `assertConfirmation` ?

File: browser-tests/utils/sharedTests.ts [hunk 9]
```diff
@@ -0,0 +1,175 @@
+import shared from '../selectors/shared';
```
Comment (santtul) [inline browser-tests/utils/sharedTests.ts:1]: should we move this file to `browser-tests/sharedTests` it's not really an util?

File: README.md [hunk 10]
```diff
@@ -54,6 +54,20 @@ To run tests:
 $ yarn test
 ```
 
+## Browser tests
+
+Browser tests are written in TypeScript with [TestCafe](https://devexpress.github.io/testcafe/) framework and they are run against [test environment](https://venepaikka-admin.test.kuva.hel.ninja) in CI as Travis Cron Job (daily) with Chrome (headless mode).
```
Comment (santtul) [inline README.md:59]: test env URL wrong

File: browser-tests/utils/sharedTests.ts [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+    .typeText(boatLength, testData.boatLength)
+    .typeText(boatName, testData.boatName)
+    .typeText(boatModel, testData.boatModel);
+
+  await t.click(nextButton);
+};
+
+export const applicantInformation = async (
+  t: TestController,
+  testData: {
```
Comment (santtul) [inline browser-tests/utils/sharedTests.ts:72]: should we create a proper types for test data(s)?

File: browser-tests/frontPage.ts (hunk 5 above)
Comment (TatuArvela) [reply browser-tests/frontPage.ts:90]: Maybe. It's simple enough to do, and the privacy policy and accessibility policy are legal responsibilities, so on the other hand why not? I'm not sure if the unit tests would catch if these are unknowingly removed.

File: browser-tests/unmarkedWinterStorage.ts (hunk 6 above)
Comment (TatuArvela) [reply browser-tests/unmarkedWinterStorage.ts:56]: I named these according to the step naming, but I'll rename them according to actions the user is performing, so yeah, `selectArea`

File: README.md (hunk 10 above)
Comment (TatuArvela) [reply README.md:59]: Copy-paste fumble, good catch!

File: browser-tests/utils/sharedTests.ts (hunk 11 above)
Comment (TatuArvela) [reply browser-tests/utils/sharedTests.ts:77]: Maybe 🤔

Comment (TatuArvela) [reply browser-tests/utils/sharedTests.ts]: Extracted the types

File: browser-tests/frontPage.ts (hunk 5 above)
Comment (santtul) [reply browser-tests/frontPage.ts:90]: Fair enough.

**Review by santtul — APPROVED** [review]
Great work! 🏅

**Pull request merged**